File: LeetCodeProblemsGo/LeetCode1578MinimumTimeToMakeRopeColorful/LeetCode1578MinimumTimeToMakeRopeColorful.go
package LeetCode1578MinimumTimeToMakeRopeColorful

func minCost(colors string, neededTime []int) int {
	previousColor := byte(0)
	totalTime := 0
	maxTime := 0
	totalBatchTime := 0
	length := len(colors)
	for i := 0; i < length; i++ {
		if colors[i] == previousColor {
			maxTime = max(maxTime, neededTime[i])
			totalBatchTime += neededTime[i]
			continue
		}
		if totalBatchTime > maxTime {
			totalTime += totalBatchTime - maxTime
		}
		previousColor = colors[i]
		maxTime = neededTime[i]
		totalBatchTime = neededTime[i]
	}
	if totalBatchTime > maxTime {
		totalTime += totalBatchTime - maxTime
	}
	return totalTime
}

File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode1578MinimumTimeToMakeRopeColorful.cpp
#include "LeetCode1578MinimumTimeToMakeRopeColorful.h"
using namespace std;

int LeetCode1578MinimumTimeToMakeRopeColorful::minCost(string colors, vector<int>& neededTime)
{
    int length = colors.length();
    char previousAmount{0};
    int batchTotalWaitTime{0};
    int maxWaitTime{0};
    int totalWaitTime{0};
    for (int i = 0; i < length; ++i)
    {
        if (colors[i] == previousAmount)
        {
            maxWaitTime = max(maxWaitTime, neededTime[i]);
            batchTotalWaitTime += neededTime[i];
            continue;
        }
        if (maxWaitTime < batchTotalWaitTime)
        {
            totalWaitTime += (batchTotalWaitTime - maxWaitTime);
        }
        previousAmount = colors[i];
        batchTotalWaitTime = neededTime[i];
        maxWaitTime = neededTime[i];
    }
    if (maxWaitTime < batchTotalWaitTime)
    {
        totalWaitTime += (batchTotalWaitTime - maxWaitTime);
    }
    return totalWaitTime;
}
